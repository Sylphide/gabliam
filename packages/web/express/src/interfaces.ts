import { express } from './express';
import { defaultMethods } from '@gabliam/web-core';

/**
 * Represent a method that create an express router
 */
export type RouterCreator = () => express.Router;

export type ExpressMethods = Extract<keyof express.Router, defaultMethods>;
