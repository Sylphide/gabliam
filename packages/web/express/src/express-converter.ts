import {
  Container,
  InjectContainer,
  INJECT_CONTAINER_KEY,
  Service,
  toPromise,
} from '@gabliam/core';
import {
  createInterceptorResolver,
  extractParameters,
  getParameterMetadata,
  isInterceptor,
  getContext,
} from '@gabliam/web-core';
import { express } from './express';

/**
 * Class for convert interceptor to Express Middleware
 */
@InjectContainer()
@Service()
export class ExpressConverter {
  public interceptorToMiddleware(clazz: any): express.RequestHandler {
    const container = <Container>(this as any)[INJECT_CONTAINER_KEY];
    const instance = createInterceptorResolver(container)(clazz);
    if (!isInterceptor(instance)) {
      throw new Error();
    }

    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const args = extractParameters(
        instance,
        'intercept',
        null,
        getContext(req),
        next,
        getParameterMetadata(instance, 'intercept')
      );

      try {
        await toPromise((instance['intercept'] as any)(...args));
        next();
      } catch (err) {
        next(err);
      }
    };
  }
}
