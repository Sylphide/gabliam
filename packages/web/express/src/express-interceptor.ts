import {
  AfterResponseInterceptor,
  Interceptor,
  InterceptorMethod,
  InterceptorConstructor,
  UseInterceptors,
} from '@gabliam/web-core';
import { METADATA_KEY } from './constants';
import { express } from './express';
import { Service } from '@gabliam/core';

/**
 * Test if target is an express interceptor
 * @param target any
 */
export const isExpressInterceptor = (target: any) =>
  Reflect.getOwnMetadata(
    METADATA_KEY.expressInterceptor,
    target.constructor || target
  ) === true;

/**
 * ExpressInterceptor decorator
 * class is an express interceptor.
 * Must return express middleware
 */
const ExpressInterceptor = () => (target: any) => {
  Reflect.defineMetadata(METADATA_KEY.expressInterceptor, true, target);
};

/**
 * Convert a Express middleware to an express interceptor
 */
export const toInterceptor = (
  mid: express.RequestHandler | express.ErrorRequestHandler,
  type: InterceptorMethod = 'intercept'
): InterceptorConstructor => {
  let clazz: InterceptorConstructor;
  if (type === 'intercept') {
    clazz = class implements Interceptor {
      intercept() {
        return mid;
      }
    };
  } else {
    clazz = class implements AfterResponseInterceptor {
      afterResponse() {
        return mid;
      }
    };
  }
  Service()(clazz);
  ExpressInterceptor()(clazz);
  return clazz;
};

/**
 * Alias for evict to use: UseInterceptors(toInterceptor(expressMid))
 */
export const UseExpressInterceptors = (
  ...mids: (express.RequestHandler | express.ErrorRequestHandler)[]
) => UseInterceptors(...mids.map(m => toInterceptor(m)));
