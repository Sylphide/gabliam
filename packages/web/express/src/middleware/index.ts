export * from './add-context';
export * from './add-middlewares';
