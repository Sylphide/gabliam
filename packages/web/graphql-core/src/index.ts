export * from './interfaces';
export { TYPE } from './constants';
export * from './graphql-core-plugin';
export * from './decorator';
export * from 'graphql-tools';
export * from 'graphql';
