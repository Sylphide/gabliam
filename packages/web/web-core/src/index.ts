export * from './constants';
export * from './cookies';
export * from './decorators';
export * from './execution-context';
export * from './gab-context';
export * from './gab-request';
export * from './gab-response';
export * from './plugin-config';
export * from './response-entity';
export * from './utils';
export * from './validate-interceptor';
export * from './validate-request';
export * from './web-configuration';
export * from './web-plugin-base';
