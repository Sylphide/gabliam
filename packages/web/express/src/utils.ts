import { toPromise } from '@gabliam/core';
import {
  ExecutionContext,
  extractParameters,
  getContext,
  Interceptor,
  InterceptorInfo,
} from '@gabliam/web-core';
import { express } from './express';
import { getExpressInterceptorType } from './express-interceptor';

export const validatorInterceptorToMiddleware = async (
  execCtx: ExecutionContext,
  { instance, paramList }: InterceptorInfo<Interceptor>
): Promise<express.RequestHandler> => {
  return async (
    req: express.Request,
    res: express.Response,
    next: express.NextFunction
  ) => {
    const args = extractParameters(
      instance,
      'intercept',
      execCtx,
      getContext(req),
      next,
      paramList
    );

    try {
      await toPromise(instance.intercept(...args));
      next();
    } catch (err) {
      next(err);
    }
  };
};

export const convertExpressInterceptorToMiddleware = async (
  expressInterceptors: InterceptorInfo<Interceptor>[]
) => {
  const requestHandlers: express.RequestHandler[] = [];
  const errorRequestHandler: express.ErrorRequestHandler[] = [];
  for (const { instance } of expressInterceptors) {
    const type = getExpressInterceptorType(instance);
    if (type === 'RequestHandler') {
      const res = await toPromise<
        express.RequestHandler | express.RequestHandler[]
      >(instance['intercept']());

      if (res === undefined) {
        continue;
      }

      if (!Array.isArray(res)) {
        requestHandlers.push(res);
      } else {
        requestHandlers.push(...res);
      }
    } else {
      const res = await toPromise<
        express.ErrorRequestHandler | express.ErrorRequestHandler[]
      >(instance['intercept']());

      if (res === undefined) {
        continue;
      }

      if (!Array.isArray(res)) {
        errorRequestHandler.unshift(res);
      } else {
        errorRequestHandler.unshift(...res);
      }
    }
  }

  return {
    requestHandlers,
    errorRequestHandler,
  };
};
