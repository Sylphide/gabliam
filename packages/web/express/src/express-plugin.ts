import {
  Container,
  GabliamPlugin,
  Plugin,
  Registry,
  Scan,
  toPromise,
} from '@gabliam/core';
import {
  APP,
  cleanPath,
  compose,
  ExecutionContext,
  getContext,
  InterceptorInfo,
  RestMetadata,
  SERVER,
  WebConfiguration,
  WebPluginBase,
  WebPluginConfig,
  WEB_PLUGIN_CONFIG,
} from '@gabliam/web-core';
import * as d from 'debug';
import * as http from 'http';
import { CUSTOM_ROUTER_CREATOR } from './constants';
import { converterValue, send } from './converter-value';
import { express } from './express';
import { isValidInterceptor } from './express-interceptor';
import { ExpressMethods, RouterCreator } from './interfaces';
import { addContextMiddleware, addMiddlewares } from './middleware';

const debug = d('Gabliam:Plugin:ExpressPlugin');

@Plugin('ExpressPlugin')
@Scan()
export class ExpressPlugin extends WebPluginBase implements GabliamPlugin {
  bindApp(
    container: Container,
    registry: Registry,
    webConfiguration: WebConfiguration
  ): void {
    container.bind(APP).toConstantValue(express());
    webConfiguration.addwebConfig({
      instance: addMiddlewares,
      order: -2,
    });

    webConfiguration.addwebConfig({
      instance: addContextMiddleware,
      order: -1,
    });
  }

  async destroy(container: Container, registry: Registry) {
    await this.stop(container, registry);
  }

  async stop(container: Container, registry: Registry) {
    try {
      // server can be undefined (if start is not called)
      const server = container.get<http.Server>(SERVER);
      return new Promise<void>(resolve => {
        server.close(() => resolve());
      });
    } catch (e) {}
  }

  async start(container: Container, registry: Registry) {
    const restConfig = container.get<WebPluginConfig>(WEB_PLUGIN_CONFIG);
    const app = container.get<express.Application>(APP);
    const port = restConfig.port;
    app.set('port', port);

    const server = http.createServer(<any>app);
    server.listen(port, restConfig.hostname);
    server.on('error', onError);
    server.on('listening', onListening);
    container.bind(SERVER).toConstantValue(server);

    /* istanbul ignore next */
    function onError(error: NodeJS.ErrnoException): void {
      // tslint:disable-next-line:curly
      if (error.syscall !== 'listen') throw error;
      const bind = typeof port === 'string' ? 'Pipe ' + port : 'Port ' + port;
      switch (error.code) {
        case 'EACCES':
          console.error(`${bind} requires elevated privileges`);
          process.exit(1);
          break;
        case 'EADDRINUSE':
          console.error(`${bind} is already in use`);
          process.exit(1);
          break;
        default:
          throw error;
      }
    }

    /* istanbul ignore next */
    function onListening(): void {
      const addr = server.address();
      const bind =
        typeof addr === 'string' ? `pipe ${addr}` : `port ${addr.port}`;
      console.log(`Listening on ${bind}`);
    }
  }

  async buildControllers(
    restMetadata: RestMetadata<ExpressMethods>,
    container: Container
  ) {
    const app = container.get<express.Application>(APP);

    // get the router creator
    let routerCreator: RouterCreator = () => express.Router();
    try {
      routerCreator = container.get<RouterCreator>(CUSTOM_ROUTER_CREATOR);
    } catch (e) {}

    for (const [
      controllerId,
      { methods, controllerPath },
    ] of restMetadata.controllerInfo) {
      const controller = container.get<object>(controllerId);
      const router = routerCreator();
      const routerPath = cleanPath(`${restMetadata.rootPath}${controllerPath}`);

      debug(`New route : "${routerPath}"`);

      for (const methodInfo of methods) {
        const execCtx = new ExecutionContext(controller, methodInfo);

        const interceptors = methodInfo.interceptors.filter(i =>
          isValidInterceptor(i)
        );

        // create handler
        const handler = this.handlerFactory(execCtx, interceptors);

        // register handler in router
        router[methodInfo.method](methodInfo.methodPath, handler);
      }
      app.use(routerPath, router);
    }
  }

  private handlerFactory(
    execCtx: ExecutionContext,
    interceptors: InterceptorInfo[]
  ): express.RequestHandler {
    return async (
      req: express.Request,
      res: express.Response,
      next: express.NextFunction
    ) => {
      const composeInterceptor = compose(
        interceptors,
        converterValue
      );
      const ctx = getContext(req);
      const methodInfo = execCtx.getMethodInfo();
      const controller = execCtx.getClass();

      let expressNextCall = false;
      const expressNext = () => {
        expressNextCall = true;
        next();
      };

      const callNext = async () => {
        const args = methodInfo.extractArgs(ctx, execCtx, expressNext);
        return await toPromise(controller[methodInfo.methodName](...args));
      };

      try {
        await composeInterceptor(ctx, execCtx, callNext);
        if (!expressNextCall) {
          send(ctx, res, methodInfo.json);
        }
      } catch (err) {
        next(err);
      }
    };
  }
}
