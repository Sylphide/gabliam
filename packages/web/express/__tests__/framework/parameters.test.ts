import {
  Controller,
  Get,
  RequestParam,
  Request,
  Response,
  QueryParam,
  Post,
  RequestBody,
  RequestHeaders,
  Cookies,
  Next,
  GabRequest,
  GabResponse,
  WebConfig,
} from '@gabliam/web-core';
import { ExpressPluginTest } from '../express-plugin-test';
import { express as e } from '../../src';
import * as supertest from 'supertest';
import * as sinon from 'sinon';
import { Config } from '@gabliam/core';

let appTest: ExpressPluginTest;

beforeEach(async () => {
  appTest = new ExpressPluginTest();
});

afterEach(async () => {
  await appTest.destroy();
});

describe('Parameters:', () => {
  test('bug parameter when value is number and passed 0', async () => {
    @Controller('/')
    class TestController {
      @Get(':id')
      public getTest(@RequestParam('id') id: number) {
        return id;
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/0')
      .expect(200);

    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the url parameter of the web request', async () => {
    @Controller('/')
    class TestController {
      @Get(':id')
      public getTest(@RequestParam('id') id: string) {
        return id;
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/foo')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the url parameter of the web request and cast to number', async () => {
    @Controller('/')
    class TestController {
      @Get(':id')
      public getTest(@RequestParam('id') id: number) {
        return [typeof id, id];
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/42')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the request object', async () => {
    @Controller('/')
    class TestController {
      @Get(':id')
      public getTest(@Request() req: GabRequest) {
        return req.params.id;
      }
    }
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/GET')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the response object', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@Response() res: GabResponse) {
        return res.originalResponse.send('foo');
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to a query parameter', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@QueryParam('id') id: string) {
        return id;
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .query('id=lolilol')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to a query parameter and cast to number', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@QueryParam('id') id: number) {
        return [typeof id, id];
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .query('id=12')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the request body', async () => {
    @Controller('/')
    class TestController {
      @Post('/')
      public getTest(@RequestBody() reqBody: string) {
        return reqBody;
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .post('/')
      .send({ foo: 'bar' })
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the request headers', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@RequestHeaders('testhead') headers: any) {
        return headers;
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .set('TestHead', 'fooTestHead')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to a cookie', async () => {
    @Controller('/')
    class TestController {
      @Get('/')
      public getCookie(@Cookies('cookie') cookie: any) {
        if (cookie) {
          return cookie;
        } else {
          return ':(';
        }
      }
    }

    @Config()
    class ServiceConfig {
      @WebConfig()
      ServerConfig(app: e.Application) {
        app.use((req, res, nextFunc) => {
          res.cookie('cookie', 'hey');
          nextFunc();
        });
      }
    }

    appTest.addClass(ServiceConfig);
    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .expect(200);
    expect(response).toMatchSnapshot();
  });

  test('should bind a method parameter to the next function', async () => {
    const spy = sinon.spy();

    @Controller('/')
    class TestController {
      @Get('/')
      public getTest(@Next() nextFunc: any) {
        spy();
        return nextFunc();
      }

      @Get('/')
      public getResult() {
        return 'foo';
      }
    }

    appTest.addClass(TestController);
    await appTest.build();
    const response = await supertest(appTest.app)
      .get('/')
      .expect(200);
    expect(spy.calledOnce).toBe(true);
    expect(response).toMatchSnapshot();
  });
});
