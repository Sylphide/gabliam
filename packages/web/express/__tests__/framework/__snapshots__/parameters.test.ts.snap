// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parameters: bug parameter when value is number and passed 0 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "1",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"1-tlifxqsNyCzxIJnRwtQKuZToQQw\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "0",
}
`;

exports[`Parameters: should bind a method parameter to a cookie 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "2",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"2-ZP5c6gUPmMfHXLaW+alufW5DcRc\\"",
    "set-cookie": Array [
      "cookie=hey; Path=/",
    ],
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": ":(",
}
`;

exports[`Parameters: should bind a method parameter to a query parameter 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "7",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"7-5Uyg9FH6Z639wlnjohqGsan13Gc\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "lolilol",
}
`;

exports[`Parameters: should bind a method parameter to a query parameter and cast to number 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-+eIvcEC3YUszYERL17BmmrIVMVw\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "[\\"number\\",12]",
}
`;

exports[`Parameters: should bind a method parameter to the next function 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-C+7Hteo/D9vJXQ3UfzxbwnXaijM\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "foo",
}
`;

exports[`Parameters: should bind a method parameter to the request body 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-pedE0BZFQNM7HX6mFsKPL6l+dUo\\"",
  },
  "req": Object {
    "data": Object {
      "foo": "bar",
    },
    "headers": Object {
      "content-type": "application/json",
    },
    "method": "POST",
  },
  "status": 200,
  "text": "{\\"foo\\":\\"bar\\"}",
}
`;

exports[`Parameters: should bind a method parameter to the request headers 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "11",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"b-zaZ1r0YCMBDq+OjlsZnp2JpqfDA\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {
      "testhead": "fooTestHead",
    },
    "method": "GET",
  },
  "status": 200,
  "text": "fooTestHead",
}
`;

exports[`Parameters: should bind a method parameter to the request object 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-8DC7vTKWbN5BA3uYqIScRrduS8E\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "GET",
}
`;

exports[`Parameters: should bind a method parameter to the response object 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-C+7Hteo/D9vJXQ3UfzxbwnXaijM\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "foo",
}
`;

exports[`Parameters: should bind a method parameter to the url parameter of the web request 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "3",
    "content-type": "text/html; charset=utf-8",
    "etag": "W/\\"3-C+7Hteo/D9vJXQ3UfzxbwnXaijM\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "foo",
}
`;

exports[`Parameters: should bind a method parameter to the url parameter of the web request and cast to number 1`] = `
Object {
  "header": Object {
    "connection": "close",
    "content-length": "13",
    "content-type": "application/json; charset=utf-8",
    "etag": "W/\\"d-f7JRWtNSw/8B+r0iQq6BpCFihJo\\"",
  },
  "req": Object {
    "data": undefined,
    "headers": Object {},
    "method": "GET",
  },
  "status": 200,
  "text": "[\\"number\\",42]",
}
`;
